src/Relude/Extra/Foldable1.hs

```diff
@@ -1,54 +1,116 @@
+
+    >>> fold1 (4 :| [5, 10 :: SG.Product Int])
+    Product {getProduct = 200}
+    -}
     fold1 :: Semigroup m => t m -> m
     fold1 = foldMap1 id
 
+    {- | Convert a non-empty data structre to a NonEmpty list.
+
+    >>> toNonEmpty (Identity 2)
+    2 :| []
+    -}
     toNonEmpty :: t a -> NonEmpty a
     toNonEmpty = foldMap1 (:|[])
 
+    {- | The first element of a non-empty data structure.
+
+    >>> head1 (1 :| [2, 3, 4])
+    1
+    -}
     head1 :: t a -> a
-    head1 = getFirst . foldMap1 First
+    head1 = SG.getFirst #. foldMap1 SG.First
 
+    {- | The last element of a non-empty data structure.
+
+    >>> last1 (1 :| [2, 3, 4])
+    4
+    -}
     last1 :: t a -> a
-    last1 = getLast . foldMap1 Last
+    last1 = SG.getLast #. foldMap1 SG.Last
 
-    maximum1 :: Ord a => t a -> a
-    maximum1 = getMax . foldMap1 Max
+    {- | The largest element of a non-empty data structure.
 
-    minimum1 :: Ord a => t a -> a
-    minimum1 = getMin . foldMap1 Min
+    >>> maximum1 (32 :| [64, 8, 128, 16])
+    128
+    -}
+    maximum1 :: Ord a => t a -> a
+    maximum1 = SG.getMax #. foldMap1 SG.Max
 
-traverse1_ :: (Foldable1 t, Applicative f) => (a -> f b) -> t a -> f ()
-traverse1_ f = foldr ((*>) . f) pass
+    {- | The smallest elemenet of a non-empty data structure.
 
-for1_ :: (Foldable1 t, Applicative f) => t a -> (a -> f b) -> f ()
-for1_ = flip traverse1_
+    >>> minimum1 (32 :| [64, 8, 128, 16])
+    8
+    -}
+    minimum1 :: Ord a => t a -> a
+    minimum1 = SG.getMin #. foldMap1 SG.Min
 
 instance Foldable1 NonEmpty where
+    foldMap1 :: Semigroup m => (a -> m) -> NonEmpty a -> m
     foldMap1 f (a :| [])     = f a
     foldMap1 f (a :| b : bs) = f a <> foldMap1 f (b :| bs)
+
+    toNonEmpty :: NonEmpty a -> NonEmpty a
     toNonEmpty = id
 
+    head1 :: NonEmpty a -> a
+    head1 = head
+
+    last1 :: NonEmpty a -> a
+    last1 = last
+
 instance Foldable1 Identity where
-    foldMap1 f = f . runIdentity
+    foldMap1 :: Semigroup m => (a -> m) -> Identity a -> m
+    foldMap1 f = f . coerce
+
+    head1 :: Identity a -> a
+    head1 = coerce
+
+    last1 :: Identity a -> a
+    last1 = coerce
+
+    maximum1 :: Ord a => Identity a -> a
+    maximum1 = coerce
+
+    minimum1 :: Ord a => Identity a -> a
+    minimum1 = coerce
 
 instance (Foldable1 f, Foldable1 g) => Foldable1 (Compose f g) where
+    foldMap1 :: Semigroup m => (a -> m) -> Compose f g a -> m
     foldMap1 f = foldMap1 (foldMap1 f) . getCompose
+
+instance (Foldable1 f, Foldable1 g) => Foldable1 (Product f g) where
+    foldMap1 :: Semigroup m => (a -> m) -> Product f g a -> m
+    foldMap1 f (Pair a b) = foldMap1 f a <> foldMap1 f b
+
+instance (Foldable1 f, Foldable1 g) => Foldable1 (Sum f g) where
+    foldMap1 :: Semigroup m => (a -> m) -> Sum f g a -> m
+    foldMap1 f (InL x) = foldMap1 f x
+    foldMap1 f (InR y) = foldMap1 f y
+
+(#.) :: Coercible b c => (b -> c) -> (a -> b) -> (a -> c)
+(#.) _f = coerce
```

Comment:
I actually thinking on adding this function to our `Relude.Extra.Newtype` module 🤔 

This might be extremely useful!